Adding plotting features for 'a priori' and 'better' analogues

PTierz opened the pull request:
### Summary

Visualising the values of single-criterion and total analogy, as well as the percentage of _better analogues_, for each _a priori_ analogue volcano selected by the user can be a valuable addition to the PyVOLCANS functionalities (please see Figs. 6, 7, in [Tierz et al., 2019](https://doi.org/10.1007/s00445-019-1336-3) for more details). The changes implemented in this branch add such functionalities.

### Description

A new function (`plot_bar_apriori_analogues()`) has been created. This function produces two different bar plots: (1) values of single-criterion (bar colours) and total analogy (total bar height) between the target volcano and each _a priori analogue_ provided; and (2) values of percentage of _better analogues_ (i.e. those with higher values of total analogy) that exist in the GVP database for the target volcano, compared to each of the _a priori analogues_ provided.

The optional flag `--plot_apriori` (or `-pa`) allows the user to produce the aforementioned plots. The optional flag `--save_figures` (or `-S`, capital S) allows the user to save the figures in png format, 600 dpi resolution.

### Usage

The following command prints the PyVOLCANS outputs to the terminal and, additionally, generates the two aforementioned figures.

```
$ pyvolcans Fuego --apriori Villarrica Llaima Pacaya Reventador Tungurahua --plot_apriori
Top 10 analogue volcanoes for Fuego, Guatemala (342090):
         name           country  smithsonian_id  total_analogy
 Klyuchevskoy            Russia          300260       0.970910
       Semeru         Indonesia          263300       0.963634
       Osorno             Chile          358010       0.955407
      Merbabu         Indonesia          263240       0.953673
       Tacana  Mexico-Guatemala          341130       0.951295
  Chikurachki            Russia          290360       0.951209
       Pavlof     United States          312030       0.950073
        Baker     United States          321010       0.949815
   Acatenango         Guatemala          342080       0.949274
   Shishaldin     United States          311360       0.948061


According to PyVOLCANS, the following percentage of volcanoes in the GVP database
are better analogues to Fuego than the 'a priori' analogues reported below:

Villarrica (357120): 2%

Llaima (357110): 2%

Pacaya (342110): 8%

Reventador (352010): 7%

Tungurahua (352080): 7%
```

The following command generates the two figures and saves them in the current working directory:

```
$ pyvolcans Fuego --apriori Villarrica Llaima Pacaya Reventador Tungurahua --plot_apriori --save_figures
```

Please note that the filename for the figures contains the weighting scheme used to run PyVOLCANS. To implement this functionality, a new function (`get_formatted_weights_text()`) has been created, which formats the weighting scheme into a machine-readable string which is used to write the filenames. In the current version, this function is also used to format the filename of the output csv file that is generated when the optional flag `--write_csv` (or `-w`) is selected.

### Acceptance criteria

+ [x] `pyvolcans Fuego --apriori` or `pyvolcans Fuego --apriori --plot_apriori` or `pyvolcans Fuego --apriori --plot_apriori --save_figures` generates the standard PyVOLCANS outputs but nothing else.
+ [x] `pyvolcans Fuego --apriori Villarrica Llaima Pacaya Reventador Tungurahua` does not produce any plot.
+ [x] `pyvolcans Fuego --apriori Villarrica Llaima Pacaya Reventador Tungurahua --plot_apriori` creates the two bar plots described above, but does not save them into png files.
+ [x] `pyvolcans Fuego --apriori Villarrica Llaima Pacaya Reventador Tungurahua --plot_apriori --save_figures` creates the two bar plots and saves the figures to png files, whose filenames end in: `*_Ts0200G0200M0200Sz0200St0200.png`.
+ [x] `pyvolcans Fuego -Ts 0.117 -G 0.253 -M 0.239 -Sz 0.032 -St 0.359 --apriori Villarrica Llaima Pacaya Reventador Tungurahua --plot_apriori --save_figures --write_csv` generates the bar plots, saves them in png format and also generates a csv file with the top 10 analogues to Fuego, for the selected weighting scheme. The csv and png files all have filenames that contain the string `*Ts0117G0253M0239Sz0032St0359.*`

mobiuscreek commented on pyvolcans/pyvolcans_func.py:
```diff
@@ -617,6 +645,91 @@ def match_name(volcano_name):
+                       ['name','ATs','AG','AM','ASz','ASt']]
+
+    # plot single- and total-analogy values for all a priori analogues
+    all_my_apriori_analogies.plot.bar(x="name",
```
This line should be returned by the first function.

mobiuscreek commented on pyvolcans/pyvolcans_func.py:
```diff
@@ -617,6 +645,91 @@ def match_name(volcano_name):
+
+    # open new figure for the 'better analogues' bar plot
+    fig2 = plt.figure()
+    plt.bar(range(len(better_analogues)), list(better_analogues.values()),
```
This line should be returned by the second function.

mobiuscreek reviewed (changes requested):
Very good additions @PTierz. Although it works it needs some finishing touches! As it stands, the plotting function doesn't have a test and we can't test it because it doesn't return anything.  To fix that I suggest some improvements below:

1. Split the plotting in two separate functions, one for the first plot and one for the second.
2. Return the plots for testing (I've marked what should be returned).

We should then be able to write a test for each and simplify the function you have. 

One a side note, for the first figure you make use of the dataframe's plotting ability that comes from pandas and uses matplotlib by default and for the second you use matplotlib directly. Bare in mind that if you have a dataframe you can use different plotting [libraries](https://pandas.pydata.org/docs/dev/development/extending.html#plotting-backends) (bokeh, altair, plotly) that could have interesting results.

mobiuscreek commented:
I hope the suggestions make sense!

PTierz commented on pyvolcans/pyvolcans_func.py:
```diff
@@ -617,6 +645,133 @@ def match_name(volcano_name):
+    my_apriori_analogues_plot = \
+        all_my_apriori_analogies.plot.bar(x="name",
+                                          y=["ATs","AG","AM","ASz","ASt"],
+                                          stacked=True)
```
Hi @mobiuscreek!

Many thanks for your comments. Could this code above help solve [this issue](https://github.com/BritishGeologicalSurvey/pyvolcans/pull/5#discussion_r619695803)?

All the best

PTierz commented on pyvolcans/pyvolcans_func.py:
```diff
@@ -617,6 +645,133 @@ def match_name(volcano_name):
+    fig2 = plt.figure()
+    my_better_analogues_plot = \
+        plt.bar(range(len(better_analogues)), list(better_analogues.values()),
+                align='center', color='blue', width=0.6)
```
Hi @mobiuscreek!

Many thanks for your comments. Could this code above help solve [this issue](https://github.com/BritishGeologicalSurvey/pyvolcans/pull/5#discussion_r619695870)?

All the best

mobiuscreek commented:
Hi @PTierz , this should work now but you should verify it before we merge. I changed the `plot_bar_better_analogues` function to use the pandas api instead of matplotlib and they now return a dataframe instead of the plot. The tests are against the dataframes that we return from each function now. I didn't change the docstrings to reflect the changes however so this is a minor change that's needed!

I also removed the `get_formatted_weights_text` function and added two lines that do the same in  `pyvolcans.py` file. The cli function starts to become too big and it would be good if we can move things like formatting functions in a separate module/file so they have their own home. It will make it cleaner especially if we keep adding new functionality but maybe this is for a later stage.

PTierz commented:
> Hi @PTierz , this should work now but you should verify it before we merge. I changed the `plot_bar_better_analogues` function to use the pandas api instead of matplotlib and they now return a dataframe instead of the plot. The tests are against the dataframes that we return from each function now. I didn't change the docstrings to reflect the changes however so this is a minor change that's needed!
> 
> I also removed the `get_formatted_weights_text` function and added two lines that do the same in `pyvolcans.py` file. The cli function starts to become too big and it would be good if we can move things like formatting functions in a separate module/file so they have their own home. It will make it cleaner especially if we keep adding new functionality but maybe this is for a later stage.

Hi @mobiuscreek!

Thank you very much for all your work!

I have implemented some further changes, pushed them and it looks like the CI workflow run was successful.

Perhaps we can wait @volcan01010 to double-check everything and then proceed with the merging?

Very many thanks both @mobiuscreek , @volcan01010!!

All the best,
Pablo
